eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -161,6 +163,14 @@ object NodeRelay {
       .modify(_.includeLocalChannelCost).setTo(true)
   }
 
+  /** If we fail to relay a payment, we may want to attempt on-the-fly funding if it makes sense. */
+  private def shouldAttemptOnTheFlyFunding(nodeParams: NodeParams, recipientFeatures_opt: Option[Features[InitFeature]], failures: Seq[PaymentFailure]): Boolean = {
+    val featureOk = Features.canUseFeature(nodeParams.features.initFeatures(), recipientFeatures_opt.getOrElse(Features.empty), Features.OnTheFlyFunding)
+    val balanceTooLow = failures.collectFirst { case f@LocalFailure(_, _, BalanceTooLow) => f }.nonEmpty
+    val routeNotFound = failures.collectFirst { case f@LocalFailure(_, _, RouteNotFound) => f }.nonEmpty
+    featureOk && (balanceTooLow || routeNotFound)
```

Comment:
Both the router and the next channel can return an "insufficient balance" error. It is inherently racy, but always happens in the context of splice outs due to #2912.

Also some additional logs help for debugging.

```suggestion
    val routerBalanceTooLow = failures.collectFirst { case f@LocalFailure(_, _, BalanceTooLow) => f }.nonEmpty
    val channelBalanceTooLow = failures.collectFirst { case f@LocalFailure(_, _, _: InsufficientFunds) => f }.nonEmpty
    val routeNotFound = failures.collectFirst { case f@LocalFailure(_, _, RouteNotFound) => f }.nonEmpty
    val res = featureOk && (routerBalanceTooLow || channelBalanceTooLow || routeNotFound)
    context.log.info(s"on-the-fly funding assessment: shouldAttemptPayToOpen=$res featureOk=$featureOk routerBalanceTooLow=$routerBalanceTooLow channelBalanceTooLow=$channelBalanceTooLow localFailures={}", failures.collect { case f: LocalFailure => f.t.getClass.getSimpleName }.mkString(","))
    res
```